Improve Home Manager support: profiles/suites, modules, extern, flake outputs

Pacman99 opened the pull request:
A really simple method of implementing #119. 

This relies on a feature that I added in home-manager master for module options of extraSpecialArgs and sharedModules, nix-community/home-manager#1793. I could try and get that backported to 20.09. But I thought I'd get some feedback on these changes first.

nrdxp commented:
bors try

Pacman99 commented:
Updated to match the lib changes. 

Also I dropped hmActivationPackages and this time exported homeConfigurations with a hyphen seperating hosts and users. So all host/user pairs get exported as `host-user`.

I think this is significantly simpler. I only exported the activation packages in the first place due to nix's issues with nested outputs. So separating with a hyphen allows for easy debugging. And with the flk script, it really only affects someone who manually builds the activation package and even that isn't that bad.

nrdxp commented:
This already is looking amazing, once you get it ready for review and it's approved, you can merge it yourself when your confident it's done. Good work.

bors delegate+

Pacman99 commented:
> This already is looking amazing, once you get it ready for review and it's approved, you can merge it yourself when your confident it's done. Good work.

Wow thats cool! Thanks! I think I might try to get the sharedModules and extraSpecialArgs PR backported to 20.09 in home-manager so I can get this merged sooner.

nrdxp commented:
Alternatively, I was thinking of upgrading the nixos flake to nixos-unstable for the time being, at least until the next stable release.

Pacman99 commented:
Ohh that makes things simpler. Can you do that as a separate commit? There might be other things to fix after upgrading, that doesn't make sense to include in this PR.

nrdxp commented:
Yeah, I'll try to manage it soon. I'm working on cleaning up some other things as well. Hopefully by the end of the day.

Pacman99 commented:
I dropped the commit updating the flake, so this should *theoretically* merge cleanly once the flake is updated. But I can fix any other conflicts once you upgrade.

Pacman99 commented:
I added some more defaults to the hmConfig. Although, optimally, that would not be needed with #127.

nrdxp commented:
I've update my local config to nixos-unstable and it seems to be working fine, after some fixes. However, I can't push it up yet, as it somehow broke the ci runner and all tests are failing with an exception. Once I figure it out, I'll push it.

Pacman99 commented:
I think the check for devshell on aarch64 will fail. Because nix flake check tries to evaluate it(not build) and nixUnstable is still broken for aarch64 on nixos-unstable - hydra hasn't updated the channel with the fix yet.

nrdxp commented:
Yeah I saw that, but actually every single derivation is [failing](https://hercules-ci.com/github/divnix/devos/jobs/119) due to an exception in hercules-ci's cpp code. I am gonna try reverting the update to the hercules flake and see if that helps. If not, it may be related to the update to Nix. I believe I can work around it either way, hopefully won't take too long.

nrdxp commented:
It seems fixing the ci-agent is going to be slightly more involved than I initially suspected. hercules-ci/hercules-ci-agent#291

We may have to rely on upstream to patch it before it works. I was thinking I could [use a container](https://discourse.nixos.org/t/nixos-container-with-different-nix-version/11976) as a workaround, but the containers seem to use the hosts nix-daemon, so it didn't pan out.

So the alternatives seem to be:
* to simply update the revision and run an older version on the ci-runner for now until a fix is found (will probably have to go with this one)
* let it stay broken and manually merge this and other PR's hoping a bunch of stuff doesn't break (probably not this one)
* use a virtual machine (might work but performance may suffer, I'll experiment)

Pacman99 commented:
Is the ci runner on an old revision evaluate the checks ok? I think the first option is the best so at least we can keep changing things.

nrdxp commented:
yes it seems to be [working again](https://hercules-ci.com/github/divnix/devos/jobs/123) after downgrading. Assuming all these checks pass, I'll push the changes with the `nixos-unstable` update.

nrdxp commented:
Before we merge this, be sure to rebase after I push the update, because I'm also going to make a change to the bors.toml. After merging my last PR as a squash merge, I decided I don't like it, and am going to use a regular merge commit instead. BORS pulls the toml from the PR and not the main branch so it's important you have the new one.

Pacman99 commented:
Ok thats makes sense, will do. Yeah squash merges aren't great for larger PR's.

nrdxp commented:
besides the merge conflicts, is this still blocked?

Pacman99 commented:
Yeah it requires features from home-manager master, so its blocked until devos updates. Or alternatively after 21.05, or a backport of the home-manager PR for shared modules and extraSpecialArgs

nrdxp commented:
Okay, if really would like to see this merged and I'm getting pretty annoyed at how difficult it's been to update to unstable. Time allowing, I'm going to attempt to start porting CI to Hydra tomorrow so we can close #192 and get this moving.

Pacman99 commented:
I rebased and fixed conflicts. But I can't actually test this new version, so after the update I will have to rebase again and run through any errors that may have come up after the rebase. 

Also I switched to use user@host pairs based on the pr in home-manager that adds `--flake` support to home manager: nix-community/home-manager#1856.

nrdxp reviewed (approved):
My review was pretty brief on this as I don't really have a ton of time and don't want to hold it up any longer. I've already delegated merge permission to you, so if you've tested it and believe it is ready. Please go ahead and merge it.

nrdxp commented:
bors try

Pacman99 commented:
This is ready for final review. I updated the home-manager input in this PR(first commit), and rebased with core. Checks succeed for me. I added some docs to users/README.md, so I think thats the last thing to check - @blaggacao I would appreciate your input for that, you're usually good with docs..
bors try

nrdxp commented on flake.nix:
```diff
@@ -18,7 +18,7 @@
       flake-compat.url = "github:BBBSnowball/flake-compat/pr-1";
       flake-compat.flake = false;
-      home.url = "github:nix-community/home-manager/release-20.09";
+      home.url = "github:nix-community/home-manager/master";
```
Total bikeshed: flake refs point to github default branch if not specified. Should we rely on this behavior or explicate branch names?

Pacman99 replied:
I switched to just home-manager. I say rely on this behavior. Its very intentional.

nrdxp commented on hosts/default.nix:
```diff
@@ -35,8 +35,13 @@ let
+        useGlobalPkgs = true;
+        useUserPackages = true;
+
+        extraSpecialArgs = extern.userSpecialArgs // { suites = suites.user; };
```
Don't user suites sometimes contain module definitions outside of home-managers scope? This may not neccisarily be an issue, but have you tested it thoroughly?

Pacman99 replied:
I tested it when I was working on the sharedModules PR, I was able to reference paths outside of home-manager perfectly fine. And considering the current direnv and git profiles are outside its scope and I switched the nixos user to use user suites. I think it works.

nrdxp commented on CHANGELOG.md:
```diff
@@ -43,6 +43,7 @@
 - iso: avoid systemd service startup [\#202](https://github.com/divnix/devos/pull/202)
 - hosts/devosSystem: pass modules as attrset [\#198](https://github.com/divnix/devos/pull/198)
 - doc: enact bootstrapping section [\#196](https://github.com/divnix/devos/pull/196)
+- flake: nixos -\> nixos-unstable [\#192](https://github.com/divnix/devos/pull/192)
```
why is this here? Isn't it already in core?

Pacman99 replied:
rebase error, fixed

nrdxp reviewed (approved):
I don't think any of my comments are major blockers. Overall this looks great. Good job on another great PR.

Pacman99 commented:
I'm going to go ahead and merge, this should be complete.
bors r+

nrdxp commented on lib/devos/devosSystem.nix:
```diff
@@ -73,10 +73,29 @@ lib.nixosSystem (args // {
+              home-manager.useUserPackages = lib.mkForce false;
+              home-manager.sharedModules = [
+                {
+                  home.packages = config.environment.systemPackages;
```
### hindsight review:
As it turns out, this may be a bit too aggressive. We may have to have a follow up PR to remove it, as it includes a lot of unrelated system packages and, worse, causes package conflicts in the community branch.

Pacman99 replied:
Yeah thats fine we can remove that line. It worked for me and I didn't think it would cause any issues.

nrdxp commented on hosts/default.nix:
```diff
@@ -35,8 +35,13 @@ let
+        useGlobalPkgs = true;
+        useUserPackages = true;
+
+        extraSpecialArgs = extern.userSpecialArgs // { suites = suites.user; };
```
We may need to add `inherit (config.nixpkgs) pkgs` as home-manager doesn't seem to have a `pkgs` argument. This may be a bug upstream however, as I believe home-manager _is_ supposed to pass in packages. In practice though, at least with this implementation, I had to add a `pkgs` arg to get doom-emacs to work, for example.

Pacman99 replied:
Are you sure thats the problem? Because home-manager already does that when `useGlobalPkgs` is true.

Pacman99 replied:
" using the system configuration's <literal>pkgs</literal>
        argument in Home Manager. This disables the Home Manager
        options <option>nixpkgs.*</option>"
^ from hm nixos module

nrdxp replied:
I ran into these issues while trying to merge this into community. I just pushed to merge so you should now be able to reproduce by simply checking out the latest community branch and running `flk home NixOS nixos`

nrdxp replied:
Perhaps shared modules don't get passed a packages arg, since that is how doom-emacs is entering the configuration. Not sure if this intentional by upstream or a bug. I'll search hm issue tracker to see if I can find some additional context.

Pacman99 replied:
Could be true, I tried adding pkgs to the hm section in `users/nixos` and used it to add vim to `home.packages` and that works fine. So your theory does make sense.

Pacman99 replied:
In home manager, the ways pkgs is set is using `_module.args.pkgs` in home-manager/modules/modules.nix. And I wonder if doesn't work in modules that were imported before that. But that doesn't make sense to me, since the module system should be recursive.

nrdxp replied:
I believe I've isolated the problem. Gonna fix and test locally then will link to an home-manager PR if I'm correct.

nrdxp replied:
okay, I've offered a minimally invasive solution in nix-community/home-manager#1879

nrdxp reviewed (commented):
We are gonna need a follow up, cleanup/enhancement PR. I wouldn't mind doing it, if nobody else wishes to do so. I've mentions a copule places that need fixing.

Pacman99 commented:
> We are gonna need a follow up, cleanup/enhancement PR. I wouldn't mind doing it, if nobody else wishes to do so. I've mentions a copule places that need fixing.

I likely won't be able to test the fixes since I can't reproduce the issues.
